### ISSUE TITLE
[MDEP-275] Figuring out duplicate class definitions using the Analyze goal

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-dependency-plugin#548]
**[Petter Måhlén](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=petterm)** opened **[MDEP-275](https://issues.apache.org/jira/browse/MDEP-275?redirect=false)** and commented

Hi,

I've pretty frequently run into issues where changes to the library structure of some product (that is, changing the way that classes are grouped into libraries) leads to the same classes being defined in more than one place. This can lead to system-dependent problems, because different versions of the same class are being loaded by different systems.

I was going to create a new goal for the dependency plugin to check for duplicate classes, but when I looked a bit closer at the analyze goal, it already had all the information needed to do that check as well, so I came up with some changes that add this functionality.

The intended usage is something like:

`mvn dependency:analyze -DcheckDuplicateClasses`

I get the feeling that I might want to add the ability to exclude certain packages (that I might be comfortable are safe to have duplicates of), so I added this option too:

`mvn dependency:analyze -DcheckDuplicateClasses -DexcludePrefixes="org., net.sf.cglib, javax.xml, junit."`

The output looks something like:

```
[WARNING] Duplicate class definitions found:
[WARNING]    com.shopzilla.common.data.ObjectFactory defined in: 
[WARNING]       com.shopzilla.site.url.c14n:model:jar:1.4:compile
[WARNING]       com.shopzilla.common.data:data-model-schema:jar:1.11:compile
[WARNING]    com.shopzilla.site.category.CategoryProvider defined in: 
[WARNING]       com.shopzilla.site2.sasClient:sas-client-core:jar:5.47:compile
[WARNING]       com.shopzilla.site2.service:common-web:jar:5.50:compile
```

A couple of notes:
- I was unable to get configuration (setting checkDuplicateClasses, etc.) using the pom to work, but I think that might be due to lack of understanding on my part.
- I don't fully understand the effect of calling compileProject() during unit tests, but I think it may be sufficient to call it only once for the duplicateClasses project, during setUp(). That would speed up the unit tests.
- I haven't added duplicate class definition checking to the AnalyzeReportMojo, because I wanted to get some feedback on whether this addition was felt to be valuable before spending any time on that.
- A lot of the unit test dummy code in the attached diff files needs cleaning up, but again I wanted to wait with that until hearing whether this might be useful to others.
- I made an API change in the ProjectDependencyAnalyzer interface, which might be an issue if there are other implementations than the default one. That change was only needed to support the 'exclude package' feature, which might not be super-important.

Cheers,
Petter

---

**Affects:** 2.1

**Attachments:**
- [dependency-analyzer.diff](https://issues.apache.org/jira/secure/attachment/12715131/dependency-analyzer.diff) (_42.44 kB_)
- [dependency-plugin.diff](https://issues.apache.org/jira/secure/attachment/12715179/dependency-plugin.diff) (_5.65 kB_)

12 votes, 9 watchers

**Comment by jira-importer** [comment]
**[Petter Måhlén](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=petterm)** commented

By the way, to make it easier for other people who are interested in using this feature, I have created a couple of repositories with the code. The repositories are:

http://github.com/pettermahlen/maven-dependency-analyzer-fork
http://github.com/pettermahlen/maven-dependency-plugin-fork

For instructions on how to start using it, see http://pettermahlen.com/2010/08/05/finding-duplicate-class-definitions-using-maven/.

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

Your timing couldn't have been better. We were just discussing the need for this the other day to help diagnose potential conflicts when switching from Maven 2.x to Maven 3.x. I'll take a look and get this merged in.

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

Raymond: Any chance you can write some unit/it tests for this new code? Currently it has none, and if you look at the mdep plugin, it is currently very thoroughly tested.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

See

https://svn.codehaus.org/mojo/trunk/sandbox/extra-enforcer-rules/src/main/java/org/codehaus/mojo/enforcer/rule/BanDuplicateClassesRule.java@13945
